### ISSUE TITLE
Limbus Cosmo Cleanse and Card

### ISSUE THREAD

**Issue by m241dan** [DarkstarProject/darkstar#2459]
Limbus only calling onBCNMEnter for the initiator,  so others arent getting card and cleanse removed.

**Comment by nesstea** [comment]
It's also allowing a person to do Omega/Ultima over and over if they turned in the chips.

**Comment by TeoTwawki** [comment]
Lamebus pretty much needs some overhauling to rewrite large parts of it. In one script its even calling a feature that was deprecated from lua before DSP was even a twinkle in anyone's eye and the core bits for battlefield entry recreate functionality that already existed just to unnecessarily enforce some specific instance IDs for what were already entirely separate battlefield IDs and call them special - the whole of limbus is just *bad* (even worse than the current dynamis implementation, if that tells you anything).

**Comment by Hozu** [comment]
@nesstea I never had infinite Ultima access, or infinite access to any of Temenos 1-3F - only Omega. I remember having to do some stupid hotfix that someone else wrote and I didn't understand why it worked out so I never contributed it here.

**Comment by nesstea** [comment]
@Hozu would you be willing to share/find such fix?

**Comment by Hozu** [comment]
Uh, I forget what I did, sorry. You might be able to find it in a fairly old (within past year) pull that didn't get merged.

**Comment by xdemolish** [comment]
you mean this https://github.com/DarkstarProject/darkstar/pull/1592 ?

**Comment by Hozu** [comment]
That's it, yeah.

**Comment by nesstea** [comment]
I'll test it out right now.

**Comment by nesstea** [comment]
It does stop the multiple chips from dropping but it doesn't fix the previous issue I stated above.

**Comment by Hozu** [comment]
Well yeah, I never said it did that, heh. All BCNMs that require everyone to have access doesn't check the other players - stuff like prime avatar fights. You can get Fenrir without having fought any prime avatars, for instance.

**Comment by TeoTwawki** [comment]
We have both `getParty()` and `getAlliance()` now so should all be fixable with the addition of a loop to remove what is needed from the other members..

**Comment by teschnei** [comment]
You don't lose your entry item until you enter

**Comment by TeoTwawki** [comment]
And we can't do this on entry because...?

**Comment by teschnei** [comment]
nobody tried, probably

**Comment by Hozu** [comment]
Oh yes, this is pretty obvious but there needs to be a distinction between BCNMs that require everyone to have the requirements, and the ones that don't. I don't know if that currently exists or not.

**Comment by nesstea** [comment]
@TeoTwawki  did you write any documentation for your getParty/alliance functions? I don't know what they return exactly.

**Comment by xdemolish** [comment]
they return a table of entities
iterate through as you would any other lua table to get the member e.g.
in alliance, all you would change is getParty() to getAlliance()
```
local party = player:getParty()
for _, member in party do
    member:doShit();
end;
```

**Comment by TeoTwawki** [comment]
After talking with Kj in IRC...We probably don't even need to use them, limbus just probably isn't checking for anything except the bcnm status when you enter. Which makes it even more stupid a bug than I initially thought.

You'd use getParty/getAlliance to do things to everyone when 1 person clicks something though.

**Comment by TeoTwawki** [comment]
@xdemolish yer example is off bud. Gotta check nil, varname off, and I just tried this appears it can do this without looping too even though every exampel in trunk does a loop then says `player'.

```lua
local party = player:getParty()
if (party ~= nil) then
    party:doStuff();
end
```

Maybe its the other way in trunk in case a player suddenly warps or something, I dunno. Still gotta do the nil check though.

**Comment by TeoTwawki** [comment]
For you email readers, editing that comment because my test was in a modified trunk...woops

**Comment by xdemolish** [comment]
whoops, prolly should've actually read the function before commenting (i remember something returned a table of entities, maybe it was zone:getPlayers() or something)

**Comment by Hozu** [comment]
Nah it checks if the other players have the two KIs when they try to enter. It simply doesn't remove them.

**Comment by TeoTwawki** [comment]
@Hozu  I mean at the point it moves you in, not the point where it decided you are eligible to go in. Even if that looks like a single step form a player perspective.

**Comment by Kushdr** [comment]
I think I fixed this and #2879 locally. I can submit later for review

**Comment by TeoTwawki** [comment]
open new issue if this is still happening
